src/Chainweb/Chainweb/PruneChainDatabase.hs

```diff
@@ -0,0 +1,394 @@
+
+instance BA.ByteArrayAccess a => CuckooFilterHash (GcHash a) where
+    cuckooHash (Salt s) (GcHash a) = fnv1a_bytes s $ BA.takeView a 8
+    cuckooFingerprint (Salt s) (GcHash a) = sip_bytes s $ BA.takeView a 8
+    {-# INLINE cuckooHash #-}
+    {-# INLINE cuckooFingerprint #-}
+
+gcHash :: a -> GcHash a
+gcHash = GcHash
+{-# INLINE gcHash #-}
+
+type Filter a = CuckooFilterIO 4 10 (GcHash a)
+
+mkFilter :: Natural -> IO (Filter a)
+mkFilter n = do
+    s <- randomIO
+    newCuckooFilter (Salt s) n
+
+-- | 'BlockPayloadHash'es and 'BlockOutputsHash'es are unique up to orphans
+-- (and genesis blocks). They can therefore be garbage collected during database
+-- pruning.
+--
+-- 'BlockTransactionHash's are not unique for block payload hashes. We use
+-- mark and sweep garbage collection to remove them.
+--
+payloadGcCallback
+    :: CasConstraint cas BlockPayload
+    => CasConstraint cas BlockOutputs
+    => CasConstraint cas OutputTree
+    => LogFunctionText
+    -> PayloadDb cas
+    -> (Maybe (Filter BlockPayloadHash))
+        -- ^ Set of marked payloads
+    -> (Maybe (Filter BlockTransactionsHash))
+        -- ^ Set of marked block transactions
+        --
+        -- This are shared globally (across chains) and it is possible that a
+        -- payload isn't marked but the block transactions of the payload are
+        -- marked.
+        --
+    -> Bool
+    -> BlockPayloadHash
+    -> IO ()
+payloadGcCallback logg db markedPayloads markedTrans isDelete h = do
+    casLookup pdb h >>= \case
+        Nothing -> logg Error
+            $ "While pruning database: payload not found for " <> encodeToText h
+            <> "; block deleted: " <> sshow isDelete
+            <> ". The database may be corrupted. In case of doubt it is recommended to delete the database and synchronize a new database"
+        Just payload -> if isDelete
+          then deleteBlockPayload logg db payload
+
+          else do
+             -- mark PayloadTransactionsHash as being used
+            forM_ markedPayloads $ \cf ->
+                tryInsert cf "payload hash" (gcHash $ _blockPayloadPayloadHash payload)
+            forM_ markedTrans $ \cf ->
+                tryInsert cf "transactions hash" (gcHash $ _blockPayloadTransactionsHash payload)
+  where
+    tryInsert cf k a = do
+        -- inserting a somewhat larger number (I think, it's actually 7) of
+        -- equal elements causes the filter to fail.
+        unlessM (member cf a) $
+            unlessM (insert cf a) $ error
+                $ "failed to insert item " <> k <> " in cuckoo filter"
+                <> ": while very rare this can happen. Usually it is resolve by retrying."
+
+    pdb = _transactionDbBlockPayloads $ _transactionDb db
+
+-- -------------------------------------------------------------------------- --
+-- Delete Payload
+--
+-- Payload Components are deleted in a way such that it is guaranteed that there
+-- are no dangling references.
+--
+-- /Tables:/
```

Comment:
Can you break up this list to comment `deleteBlockPayload` and `deleteBlockTransactions` separately?